src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,258 @@
+    /// Egress provider for egressing stream data to the s3 storage.
+    /// </summary>
+    internal class S3StorageEgressProvider : EgressProvider<S3StorageEgressProviderOptions>
+    {
+        public S3StorageEgressProvider(ILogger<S3StorageEgressProvider> logger) : base(logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<CancellationToken, Task<Stream>> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await using var stream = await action(token);
+                var resourceId = await UploadToStorageAsync(options, stream, artifactSettings, token);
+                Logger?.EgressProviderSavedStream(EgressProviderTypes.S3Storage, resourceId);
+                return resourceId;
+            }
+            catch (EgressException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<Stream, CancellationToken, Task> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                var client = await CreateClientAsync(options, token);
+                var uploadId = await InitMultiPartUploadAsync(client, options.BucketName, artifactSettings, token);
+                await using var stream = new MultiPartUploadStream(client, options.BucketName, artifactSettings.Name, uploadId, options.CopyBufferSize!.Value);
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await action(stream, token);
+                await stream.FinalizeAsync(token); // force to push the last part
+                stream.Close();
+
+                if (stream.Parts.Count == 0)
+                {
+                    Logger?.LogDebug(Strings.Message_EgressS3NoContent);
+                    return string.Empty; // nothing to return here
+                }
+
+                await CompleteMultiPartUploadAsync(client, options.BucketName, uploadId, artifactSettings, stream.Parts, token);
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        private async Task<string> UploadToStorageAsync(
```

Comment:
This is only called by a single method and the caller is fairly small. Inline the content of this method with the first EgressAsync method.